tests/src/test/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLogTest.java

```diff
@@ -0,0 +1,146 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop.storage;
+
+import io.streamnative.pulsar.handlers.kop.KafkaServiceConfiguration;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.RecordTooLargeException;
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.ControlRecordType;
+import org.apache.kafka.common.record.EndTransactionMarker;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.common.utils.Time;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * Unit test {@link PartitionLog}.
+ */
+@Slf4j
+public class PartitionLogTest {
+
+    private static final KafkaServiceConfiguration kafkaConfig = new KafkaServiceConfiguration();
+
+    private static final PartitionLog PARTITION_LOG = new PartitionLog(
+            kafkaConfig,
+            Time.SYSTEM,
+            new TopicPartition("test", 1),
+            "test",
+            "test",
+            null,
+            new ProducerStateManager("test"));
+
+    @DataProvider(name = "compressionTypes")
+    Object[] allCompressionTypes() {
+        return Arrays.stream(CompressionType.values()).map(x -> (Object) x).toArray();
+    }
+
+    @Test(dataProvider = "compressionTypes")
+    public void testAnalyzeAndValidateRecords(CompressionType compressionType) {
+        MemoryRecords memoryRecords = buildMemoryRecords(new int[]{3, 2, 5}, compressionType, 100);
+        PartitionLog.LogAppendInfo appendInfo =
+                PARTITION_LOG.analyzeAndValidateRecords(memoryRecords);
+        Assert.assertEquals(appendInfo.shallowCount(), 3);
+        Assert.assertEquals(appendInfo.numMessages(), 10);
+        Assert.assertEquals(appendInfo.validBytes(), memoryRecords.validBytes());
+        Assert.assertEquals(appendInfo.sourceCodec().codec(), compressionType.id);
+        Assert.assertFalse(appendInfo.isTransaction());
```

Comment:
I found all tests only validate the following fields:

```java
        private int numMessages;
        private int shallowCount;
        private boolean isTransaction;
        private int validBytes;
```

Only this test validates the `sourceCodec` field. So I think we can add a method to validate the other fields of `appendInfo` are the default values.

Reply:
Addressed.

Reply:
The `lastOffsetOfFirstBatch` field is still tested, PTAL.

Reply:
Addressed. PTAL.